---
title: "Lab_13_RNASeq_project"
author: "Ryan Bench (PID: A69038034)"
format: pdf
toc: true
---
## Background 

Today we will run through a complete RNASeq analysis

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.

## Data Import

```{r}
counts <- read.csv("GSE37704_featurecounts.csv", row.names = 1)
metadata <- read.csv("GSE37704_metadata.csv")
```

Check correspondence of `metadata` and `counts` (i.e. that the columns in `counts` match the rows of `metadata`)

```{r}
metadata
```

```{r}
colnames(counts)
```

```{r}
metadata$id
```
```{r}
counts <- counts[,-1]

test_cols <- !all(colnames(counts) == metadata$id)
```

Also lets remove low count genes

```{r}
tot.counts <- rowSums(counts)
head(tot.counts)
```
Let's remove all zero count genes

```{r}
zero.inds <- tot.counts == 0
head(zero.inds)
```

```{r}
counts <- counts[!zero.inds, ]
```


```{r}
all(colnames(counts)[-1] == metadata$id)
```

```{r}
if( !all(c(T,T,F))) {
  message("Wow..... there is a problem with the metadata counts setup!")
}
```

```{r}
head(counts)
```


## Setup for DESeq

```{r}
library(DESeq2)
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, colData = metadata, design = ~condition)
```

## Run DESeq

```{r}
dds <- DESeq(dds)
```

## Get results

```{r}
res <- results(dds)
```

```{r}
head(res)
```

## Add annotation

library()
```{r}
library(AnnotationDbi)
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

head(res, 10)
```

## Visulaize results

```{r}
library(ggplot2)
my_cols <- rep("gray", nrow(res))
my_cols[abs(res$log2FoldChange) >= 2] <- "blue"

ggplot(res) +
  aes(log2FoldChange, -log(padj)) +
  geom_point(col = my_cols) +
  geom_vline(xintercept = c(-2, 2), col = "red") +
  geom_hline(yintercept = -log(0.05), col = "red")
```


## Pathway analysis

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)
```


```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
head(keggres$less)
```

```{r, message=FALSE}
library(pathview)

pathview(gene.data=foldchanges, pathway.id = "hsa04110")
```

![](hsa04110.pathview.png)

### Let's try GO analysis and compare to KEGG analysis

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)
```

```{r}
head( gobpres$less )
```
## Reactome

Some folks really like Reactomes online (i.e. their webpage viewer) rather than the R package of the same name (available from bioconductor).

To use the website viewer (https://reactome.org/) we want to upload our set of gene symbols for the genes we want to focus on (here those with a P-value below 0.05)

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```


## Save results
```{r}
write.csv(res, file = "myresults.csv")
```


```{r}
save(res, file = "my_results.RData")
```



